File: leetcode/leetcode_easy/minStack.py
'''
 Design a stack that supports push, pop, top, 
 and retrieving the minimum element in constant time.

    push(x) -- Push element x onto stack.
    pop() -- Removes the element on top of the stack.
    top() -- Get the top element.
    getMin() -- Retrieve the minimum element in the stack.

'''
class MinStack(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.min = None


    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        self.stack.append(x)
        if self.min is None:
            self.min = x
        else:
            if self.min > x:
                self.min = x


    def pop(self):
        """
        :rtype: void
        """
        if self.min == self.stack.pop():
            if len(self.stack) == 0:
                self.min = None
            else:
                self.min = self.stack[0]
                for i in self.stack[1:]:
                    if i < self.min:
                        self.min = i


    def top(self):
        """
        :rtype: int
        """
        return self.stack[-1]



    def getMin(self):
        """
        :rtype: int
        """
        return self.min



        # Your MinStack object will be instantiated and called as such:
        # obj = MinStack()
        # obj.push(x)
        # obj.pop()
        # param_3 = obj.top()
        # param_4 = obj.getMin()

File: leetcode/leetcode_easy_java/leetcode_easy/src/shy/MinStack.java
package shy;

import java.util.ArrayList;

public class MinStack {

	/** initialize your data structure here. */
	ArrayList<Integer> arrayList = new ArrayList<Integer>();
	int min;

	public MinStack() {

	}

	public void push(int x) {
		if (arrayList.isEmpty()) {
			min = x;
			arrayList.add(x);
		} else {
			min = Math.min(min, x);
			arrayList.add(x);

		}

	}

	public void pop() {
		int tmp = arrayList.get(arrayList.size() - 1);
		arrayList.remove(arrayList.size() - 1);
		if (tmp == min) {

			if (!arrayList.isEmpty()) {
				min = arrayList.get(0);
				for (int i : arrayList) {
					min = Math.min(i, min);
				}
			}

		}

	}

	public int top() {
		return arrayList.get(arrayList.size() - 1);

	}

	public int getMin() {
		return min;

	}
}
